platform/sensor-stream-manager/sensor-stream-manager.js

```diff
@@ -1,64 +1,205 @@
+        // check the example below
+        // sensorStream = {
+        //     "sensor1-id": {
+        //         "gateway1-ip": [ "app1-topic", "app2-topic" ],
+        //         "geteway2-ip": [ "app3-topic"]
+        //     },
+        //     "sensor2-id": {
+        //         "gateway1-ip": [ "app1-topic" ],
+        //         "geteway2-ip": [ "app3-topic"]
+        //     }
+        // }
+        this.sensorStream = {};
+
+        // TODO: policy
+        this.policy = {};
+
+        // mqttClients = {
+        //     "gateway-ip": client
+        // }
+        this.mqttClients = {};
+
+        this.registerMQTTClient("localhost", "gateway-data");
+    }
+
+    /* ===== public functions ===== */
+
+    /**
+     * This function registers topic for local sensors.
+     * @param {string} ip - gateway' ip
+     * @param {string[]} sensorIds - an array of sensor id
+     * @param {string} topic
+     */
+    registerLocalGatewayTopic(ip, sensorIds, topic) {
+        for(const id of sensorIds) {
+            if(!this.sensorStream[id]) {
+                this.sensorStream[id] = {};
+            }
+            if(!this.sensorStream[id][ip]) {
+                this.sensorStream[id][ip] = [];
+            }
+            this.sensorStream[id][ip].push(topic);
+        }
+    }
+
+    /**
+     * This function sends sensor requirement to the remote gateway.
+     * @param {string} ip -remote gateway's ip
+     * @param {string[]} sensorIds - an array of sensor id
+     * @param {string} topic
+     */
+    registerRemoteGatewayTopic(ip, sensorIds, topic) {
+        // Remote gateway's register-topic url
+        const gatewayUrl = `http://${ip}:5000/gateway/register-app-sensor-reqruirement`;
+        // Request body
+        const body = {
+            ip: this.gatewayIp,
+            sensors: sensorIds,
+            topic: topic,
+        };
+        // Send application's sensor requirement to remote gateway
+        fetch(gatewayUrl, {
+            method: 'POST',
+            body: JSON.stringify(body),
+            headers: {'Content-Type': 'application/json'},
+            timeout: 5000
+        }).then(res => {
+            if(res.status === 200) {
+                console.log(`[INFO] Sent "${topic}" to ${ip} successfully!`);
+            } else {
+                console.error(`[ERROR] Failed to send "${topic}" to ${ip} with status ${res.status}.`);
+            }
+        }).catch(err => {
+            console.error(`[ERROR] Failed to send "${topic}" to ${ip}.`);
             console.error(err);
-        } else {
-            console.log(`[INFO] MQTT client subscribed to "gateway-data" topic successfully!`);
+        });
+    }
+
+    /**
+     * This function register MQTT clients.
+     * @param {string} ip - MQTT broker's ip
+     * @param {string} mqttTopic
+     */
+    registerMQTTClient(ip, mqttTopic) {
+        // Check if the MQTT client exists
+        if(ip === "localhost") ip = this.gatewayIp;
```

Comment:
Stick with same if style
if() {
...
}

Add a comment stating that we need to convert this because the metadata gives things in terms of IP addresses

Reply:
I found that client can connect to local broker with its IP, so I remove all locahost.